File: Reports/TransitionSummaryReport.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Core
Imports System.Globalization
Imports SyncroSim.Core.Forms
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class TransitionSummaryReport
    Inherits ExportTransformer

    Protected Overrides Sub Export(location As String, exportType As ExportType)
        Me.InternalExport(location, exportType, True)
    End Sub

    Friend Sub InternalExport(location As String, exportType As ExportType, showMessage As Boolean)

        Dim AmountLabel As String = Nothing
        Dim TermUnit As TerminologyUnit
        Dim dsterm As DataSheet = Me.Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        Dim columns As ExportColumnCollection = Me.CreateColumnCollection()

        GetAmountLabelTerminology(dsterm, AmountLabel, TermUnit)

        If (exportType = ExportType.ExcelFile) Then

            Dim WorksheetName As String = String.Format(CultureInfo.InvariantCulture, "{0} by Transition Group", AmountLabel)
            Me.ExcelExport(location, columns, Me.CreateReportQuery(False), WorksheetName)

        Else

            columns.Remove("ScenarioName")
            Me.CSVExport(location, columns, Me.CreateReportQuery(True))

            If (showMessage) Then
                FormsUtilities.InformationMessageBox("Data saved to '{0}'.", location)
            End If

        End If

    End Sub

    Private Function CreateColumnCollection() As ExportColumnCollection

        Dim c As New ExportColumnCollection()

        Dim AmountLabel As String = Nothing
        Dim UnitsLabel As String = Nothing
        Dim TermUnit As TerminologyUnit
        Dim PrimaryStratumLabel As String = Nothing
        Dim SecondaryStratumLabel As String = Nothing
        Dim dsterm As DataSheet = Me.Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        Dim TimestepLabel As String = GetTimestepUnits(Me.Project)

        GetAmountLabelTerminology(dsterm, AmountLabel, TermUnit)
        GetStratumLabelTerminology(dsterm, PrimaryStratumLabel, SecondaryStratumLabel)
        UnitsLabel = TerminologyUnitToString(TermUnit)

        Dim AmountTitle As String = String.Format(CultureInfo.InvariantCulture, "{0} ({1})", AmountLabel, UnitsLabel)

        c.Add(New ExportColumn("ScenarioID", "Scenario ID"))
        c.Add(New ExportColumn("ScenarioName", "Scenario"))
        c.Add(New ExportColumn("Iteration", "Iteration"))
        c.Add(New ExportColumn("Timestep", TimestepLabel))
        c.Add(New ExportColumn("Stratum", PrimaryStratumLabel))
        c.Add(New ExportColumn("SecondaryStratum", SecondaryStratumLabel))
        c.Add(New ExportColumn("TransitionGroup", "Transition Group"))
        c.Add(New ExportColumn("AgeMin", "Age Min"))
        c.Add(New ExportColumn("AgeMax", "Age Max"))
        c.Add(New ExportColumn("Amount", AmountTitle))

        c("Amount").DecimalPlaces = 2
        c("Amount").Alignment = ColumnAlignment.Right

        Return c

    End Function

    Private Function CreateReportQuery(ByVal isCSV As Boolean) As String

        Dim ScenFilter As String = Me.CreateActiveResultScenarioFilter()

        If (isCSV) Then

            Return String.Format(CultureInfo.InvariantCulture,
                "SELECT " &
                "STSim_OutputStratumTransition.ScenarioID, " &
                "STSim_OutputStratumTransition.Iteration,  " &
                "STSim_OutputStratumTransition.Timestep,  " &
                "STSim_Stratum.Name AS Stratum,  " &
                "STSim_SecondaryStratum.Name AS SecondaryStratum,  " &
                "STSim_TransitionGroup.Name as TransitionGroup, " &
                "STSim_OutputStratumTransition.AgeMin, " &
                "STSim_OutputStratumTransition.AgeMax, " &
                "STSim_OutputStratumTransition.Amount " &
                "FROM STSim_OutputStratumTransition " &
                "INNER JOIN STSim_Stratum ON STSim_Stratum.StratumID = STSim_OutputStratumTransition.StratumID " &
                "LEFT JOIN STSim_SecondaryStratum ON STSim_SecondaryStratum.SecondaryStratumID = STSim_OutputStratumTransition.SecondaryStratumID " &
                "INNER JOIN STSim_TransitionGroup ON STSim_TransitionGroup.TransitionGroupID = STSim_OutputStratumTransition.TransitionGroupID " &
                "WHERE STSim_OutputStratumTransition.ScenarioID IN ({0})  " &
                "ORDER BY " &
                "STSim_OutputStratumTransition.ScenarioID, " &
                "STSim_OutputStratumTransition.Iteration, " &
                "STSim_OutputStratumTransition.Timestep, " &
                "STSim_Stratum.Name, " &
                "STSim_SecondaryStratum.Name, " &
                "STSim_TransitionGroup.Name, " &
                "AgeMin, " &
                "AgeMax",
                ScenFilter)

        Else

            Return String.Format(CultureInfo.InvariantCulture,
                "SELECT " &
                "STSim_OutputStratumTransition.ScenarioID, " &
                "SSim_Scenario.Name AS ScenarioName,  " &
                "STSim_OutputStratumTransition.Iteration,  " &
                "STSim_OutputStratumTransition.Timestep,  " &
                "STSim_Stratum.Name AS Stratum,  " &
                "STSim_SecondaryStratum.Name AS SecondaryStratum,  " &
                "STSim_TransitionGroup.Name as TransitionGroup, " &
                "STSim_OutputStratumTransition.AgeMin, " &
                "STSim_OutputStratumTransition.AgeMax, " &
                "STSim_OutputStratumTransition.Amount " &
                "FROM STSim_OutputStratumTransition " &
                "INNER JOIN SSim_Scenario ON SSim_Scenario.ScenarioID = STSim_OutputStratumTransition.ScenarioID " &
                "INNER JOIN STSim_Stratum ON STSim_Stratum.StratumID = STSim_OutputStratumTransition.StratumID " &
                "LEFT JOIN STSim_SecondaryStratum ON STSim_SecondaryStratum.SecondaryStratumID = STSim_OutputStratumTransition.SecondaryStratumID " &
                "INNER JOIN STSim_TransitionGroup ON STSim_TransitionGroup.TransitionGroupID = STSim_OutputStratumTransition.TransitionGroupID " &
                "WHERE STSim_OutputStratumTransition.ScenarioID IN ({0})  " &
                "ORDER BY " &
                "STSim_OutputStratumTransition.ScenarioID, " &
                "SSim_Scenario.Name, " &
                "STSim_OutputStratumTransition.Iteration, " &
                "STSim_OutputStratumTransition.Timestep, " &
                "STSim_Stratum.Name, " &
                "STSim_SecondaryStratum.Name, " &
                "STSim_TransitionGroup.Name, " &
                "AgeMin, " &
                "AgeMax",
                ScenFilter)

        End If

    End Function

End Class

File: src/Reports/TransitionSummaryReport.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Core;
using SyncroSim.Core.Forms;
using System.Globalization;

namespace SyncroSim.STSim
{
    internal class TransitionSummaryReport : ExportTransformer
    {
        protected override void Export(string location, ExportType exportType)
        {
            this.InternalExport(location, exportType, true);
        }

        internal void InternalExport(string location, ExportType exportType, bool showMessage)
        {
            string AmountLabel = null;
            TerminologyUnit TermUnit = 0;
            DataSheet dsterm = this.Project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);
            ExportColumnCollection columns = this.CreateColumnCollection();

            TerminologyUtilities.GetAmountLabelTerminology(dsterm, ref AmountLabel, ref TermUnit);

            if (exportType == ExportType.ExcelFile)
            {
                string WorksheetName = string.Format(CultureInfo.InvariantCulture, "{0} by Transition Group", AmountLabel);
                this.ExportToExcel(location, columns, this.CreateReportQuery(false), WorksheetName);
            }
            else
            {
                columns.Remove("ScenarioName");
                this.ExportToCSVFile(location, columns, this.CreateReportQuery(true));

                if (showMessage)
                {
                    FormsUtilities.InformationMessageBox("Data saved to '{0}'.", location);
                }
            }
        }

        private ExportColumnCollection CreateColumnCollection()
        {
            ExportColumnCollection c = new ExportColumnCollection();

            string AmountLabel = null;
            string UnitsLabel = null;
            TerminologyUnit TermUnit = 0;
            string PrimaryStratumLabel = null;
            string SecondaryStratumLabel = null;
            string TertiaryStratumLabel = null;
            DataSheet dsterm = this.Project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);
            string TimestepLabel = TerminologyUtilities.GetTimestepUnits(this.Project);

            TerminologyUtilities.GetAmountLabelTerminology(dsterm, ref AmountLabel, ref TermUnit);
            TerminologyUtilities.GetStratumLabelTerminology(dsterm, ref PrimaryStratumLabel, ref SecondaryStratumLabel, ref TertiaryStratumLabel);
            UnitsLabel = TerminologyUtilities.TerminologyUnitToString(TermUnit);

            string AmountTitle = string.Format(CultureInfo.InvariantCulture, "{0} ({1})", AmountLabel, UnitsLabel);

            c.Add(new ExportColumn("ScenarioId", "Scenario Id"));
            c.Add(new ExportColumn("ScenarioName", "Scenario"));
            c.Add(new ExportColumn("Iteration", "Iteration"));
            c.Add(new ExportColumn("Timestep", TimestepLabel));
            c.Add(new ExportColumn("Stratum", PrimaryStratumLabel));
            c.Add(new ExportColumn("SecondaryStratum", SecondaryStratumLabel));
            c.Add(new ExportColumn("TertiaryStratum", TertiaryStratumLabel));
            c.Add(new ExportColumn("TransitionGroup", "Transition Group"));
            c.Add(new ExportColumn("AgeMin", "Age Min"));
            c.Add(new ExportColumn("AgeMax", "Age Max"));
            c.Add(new ExportColumn("Amount", AmountTitle));

            c["Amount"].DecimalPlaces = 2;
            c["Amount"].Alignment = ColumnAlignment.Right;

            return c;
        }

        private string CreateReportQuery(bool isCSV)
        {
            string ScenFilter = this.ExportCreateActiveResultScenarioFilter();

            if (isCSV)
            {
                return string.Format(CultureInfo.InvariantCulture, 
                    "SELECT " + "stsim_OutputStratumTransition.ScenarioId, " + "stsim_OutputStratumTransition.Iteration,  " + 
                    "stsim_OutputStratumTransition.Timestep,  " + "stsim_Stratum.Name AS Stratum,  " + "stsim_SecondaryStratum.Name AS SecondaryStratum,  " + 
                    "stsim_TertiaryStratum.Name AS TertiaryStratum,  " + "stsim_TransitionGroup.Name as TransitionGroup, " + "stsim_OutputStratumTransition.AgeMin, " + 
                    "stsim_OutputStratumTransition.AgeMax, " + "stsim_OutputStratumTransition.Amount " + "FROM stsim_OutputStratumTransition " + 
                    "INNER JOIN stsim_Stratum ON stsim_Stratum.StratumId = stsim_OutputStratumTransition.StratumId " +
                    "LEFT JOIN stsim_SecondaryStratum ON stsim_SecondaryStratum.SecondaryStratumId = stsim_OutputStratumTransition.SecondaryStratumId " +
                    "LEFT JOIN stsim_TertiaryStratum ON stsim_TertiaryStratum.TertiaryStratumId = stsim_OutputStratumTransition.TertiaryStratumId " + 
                    "INNER JOIN stsim_TransitionGroup ON stsim_TransitionGroup.TransitionGroupId = stsim_OutputStratumTransition.TransitionGroupId " + 
                    "WHERE stsim_OutputStratumTransition.ScenarioId IN ({0})  " + "ORDER BY " + "stsim_OutputStratumTransition.ScenarioId, " + 
                    "stsim_OutputStratumTransition.Iteration, " + "stsim_OutputStratumTransition.Timestep, " + "stsim_Stratum.Name, " + 
                    "stsim_SecondaryStratum.Name, " + "stsim_TertiaryStratum.Name, " + "stsim_TransitionGroup.Name, " + "AgeMin, " + "AgeMax", ScenFilter);
            }
            else
            {
                return string.Format(CultureInfo.InvariantCulture,
                    "SELECT " + "stsim_OutputStratumTransition.ScenarioId, " +
                    "core_Scenario.Name AS ScenarioName,  " + "stsim_OutputStratumTransition.Iteration,  " + "stsim_OutputStratumTransition.Timestep,  " + 
                    "stsim_Stratum.Name AS Stratum,  " + "stsim_SecondaryStratum.Name AS SecondaryStratum,  " + "stsim_TertiaryStratum.Name AS TertiaryStratum,  " + 
                    "stsim_TransitionGroup.Name as TransitionGroup, " + "stsim_OutputStratumTransition.AgeMin, " + "stsim_OutputStratumTransition.AgeMax, " + 
                    "stsim_OutputStratumTransition.Amount " + "FROM stsim_OutputStratumTransition " +
                    "INNER JOIN core_Scenario ON core_Scenario.ScenarioId = stsim_OutputStratumTransition.ScenarioId " + 
                    "INNER JOIN stsim_Stratum ON stsim_Stratum.StratumId = stsim_OutputStratumTransition.StratumId " + 
                    "LEFT JOIN stsim_SecondaryStratum ON stsim_SecondaryStratum.SecondaryStratumId = stsim_OutputStratumTransition.SecondaryStratumId " + 
                    "LEFT JOIN stsim_TertiaryStratum ON stsim_TertiaryStratum.TertiaryStratumId = stsim_OutputStratumTransition.TertiaryStratumId " + 
                    "INNER JOIN stsim_TransitionGroup ON stsim_TransitionGroup.TransitionGroupId = stsim_OutputStratumTransition.TransitionGroupId " +
                    "WHERE stsim_OutputStratumTransition.ScenarioId IN ({0})  " + "ORDER BY " + "stsim_OutputStratumTransition.ScenarioId, " + 
                    "core_Scenario.Name, " + "stsim_OutputStratumTransition.Iteration, " + "stsim_OutputStratumTransition.Timestep, " +
                    "stsim_Stratum.Name, " + "stsim_SecondaryStratum.Name, " + "stsim_TertiaryStratum.Name, " + "stsim_TransitionGroup.Name, " + 
                    "AgeMin, " + "AgeMax", ScenFilter);
            }
        }
    }
}
